infrastructure/smart-contracts/contracts/upgrades/esXai/esXai4.sol

```diff
@@ -0,0 +1,380 @@
+     */
+    function transferFrom(address from, address to, uint256 amount) public override returns (bool) {
+        require(_whitelist.contains(from) || _whitelist.contains(to), "Transfer not allowed: address not in whitelist");
+        return super.transferFrom(from, to, amount);
+    }
+
+    /**
+     * @dev Function to start the redemption process
+     * @param amount The amount of esXai to redeem.
+     * @param duration The duration of the redemption process in seconds.
+     */
+    function startRedemption(uint256 amount, uint256 duration) public {
+        require(_redemptionActive, "Redemption is currently inactive");
+        require(amount > 0, "Invalid Amount");
+        require(balanceOf(msg.sender) >= amount, "Insufficient esXai balance");
+        require(duration == 15 days || duration == 90 days || duration == 180 days, "Invalid duration");
+        
+        // Check if the sender failed KYC
+        bool failedKyc = PoolFactory3(poolFactoryAddress).failedKyc(msg.sender);
+        require(!failedKyc, "KYC failed, cannot redeem");
+        
+        // No longer transferring esXai from the sender's account to this contract
+        // Stoing redemption claim as voucherIssued instead
+
+        // Confirm the user has the appropriate amount of esXai available
+        uint256 currentBalance = balanceOf(msg.sender);        
+        uint256 totalesXaiStaked = PoolFactory3(poolFactoryAddress).getTotalesXaiStakedByUser(msg.sender);
+        uint256 totalesXaiPendingRedemption = _totalPendingRedemptions[msg.sender];
+
+        uint256 availableesXai = currentBalance + totalesXaiStaked - totalesXaiPendingRedemption;
+
+        require(availableesXai >= amount, "Insufficient esXai balance");
+        
+        // Increment the total pending redemptions
+        _totalPendingRedemptions[msg.sender] += amount;
+
+        // Store the redemption request
+        _extRedemptionRequests[msg.sender].push(RedemptionRequestExt(amount, block.timestamp, duration, 0, false, false, true, [uint256(0),0,0,0,0]));
+        emit RedemptionStarted(msg.sender, _extRedemptionRequests[msg.sender].length - 1);
+    }
+
+    /**
+     * @dev Function to cancel the redemption process
+     * @param index The index of the redemption request to cancel.
+     */
+    function cancelRedemption(uint256 index) public {
+        require(_redemptionActive, "Redemption is currently inactive");
+        RedemptionRequestExt storage request = _extRedemptionRequests[msg.sender][index];
+        require(request.amount > 0, "Invalid request");
+        require(!request.completed, "Redemption already completed");
+
+        // Mark the redemption request as completed
+        request.completed = true;
+        request.cancelled = true;
+        request.endTime = block.timestamp;
+
+        // Decrement the total pending redemptions
+        _totalPendingRedemptions[msg.sender] -= request.amount;
+
+        emit RedemptionCancelled(msg.sender, index);
```

Comment:
Does this not need to consider if the vouchers has been issued ?

Reply:
See my original comment on the PR. I asked this exact question.
Adding it now.